### ISSUE TITLE
Reach jarfile not using given config file.

### ISSUE THREAD

**Issue by pagreene** [clulab/reach#549]
We create a jar file to run reach (for portability), using the following in the top reach directory:
```
echo 'mainClass in assembly := Some("org.clulab.reach.RunReachCLI")' >> build.sbt
[sudo] sbt assembly
```
(Thus sudo is supposedly not needed, but sbt won't run on my system without it) We would then like to be able to run the script via
```
java -jar <name of the jarfile>.jar <config file>
```
This has worked with previous versions of reach, however, it does not work for 1.3.4 or 1.3.5. I have verified that the conf file is valid, by running (successfully in both versions)
```
[sudo] sbt -Dconfig.file=<path to config file> "runMain org.clulab.reach.RunReachCLI"
```
Note that
```
[sudo] sbt "runMain org.clulab.reach.RunReachCLI <path to config file>"
```
Shows essentially the same problem as with running with the jar file.

The actual failure occurs when the system tries to access '[Home]/reach/papers', which of course does not exist.

**Comment by hickst** [comment]
> We create a jar file to run reach (for portability), using the following in the top reach directory:
> ```
> echo 'mainClass in assembly := Some("org.clulab.reach.RunReachCLI")' >> build.sbt
> [sudo] sbt assembly
> ```

Bypassing the Reach build file is contra-indicated and discouraged. In this particular case, setting `mainClass` is not even necessary as the default `mainClass` is exactly what you want (`org.clulab.reach.RunReachCLI`). You can set another class to build by using the `-DmainClass` flag to `sbt` at build time. Please see the Wiki page on building a FAT JAR: https://github.com/clulab/reach/wiki/Using-Reach-from-a-FAT-JAR for more detail.

> (Thus sudo is supposedly not needed, but sbt won't run on my system without it)

I've never encountered this particular problem but my hunch is that `sbt` is incorrectly installed on your system or that the installation files have the wrong permissions. I would clean and re-install `sbt` or ask your system administrator to do it.

> We would then like to be able to run the script via
> ```
> java -jar <name of the jarfile>.jar <config file>
> ```
> This has worked with previous versions of reach, however, it does not work for 1.3.4 or 1.3.5.

I'm not familiar with that historical capability, having never used it, but you should be able to achieve the same thing using the `-Dconfig.file` argument. For example:

```java -Dconfig.file=/Users/hickst/temp/reach-test-app.conf -jar reach-1.3.5-SNAPSHOT-FAT.jar```

> The actual failure occurs when the system tries to access '[Home]/reach/papers', which of course does not exist.

Reach is designed to be configured before compilation using the provided configuration files, the main one of which is the `application.conf` file in `main/src/main/resources`. The `${HOME}` path is a default value which works on many OS X and Linux systems, but it is expected that users will set this variable to an existing path before compiling Reach. Please see the instructions for running, found on the Wiki, paying special attention to **Step 2**. https://github.com/clulab/reach/wiki/Running-REACH

**Comment by bgyori** [comment]
Hi Tom, thanks! Passing -Dconfig.file is working. However, unexpectedly, I found that after REACH is done reading all the papers in the input folder, it doesn't exit, and keeps waiting. Is this the expected behavior? Is there a way to configure REACH CLI so that it exits when done?

**Comment by hickst** [comment]
Yea....this is a known problem: https://github.com/clulab/reach/issues/547 resulting from the Akka server threads remaining active and preventing the JVM from exiting. I was investingating a solution but the coordination was too complicated and we concluded that we needed a larger code refactoring anyway. We anticipate that the code refactoring will eliminate the issue but it will be awhile before we can do it.

**Comment by bgyori** [comment]
Alright, do you know what would be the latest REACH commit that we could use which doesn't suffer from this issue?

**Comment by hickst** [comment]
Unfortunately, it was awhile ago.....late July....and I wouldn't recommend dropping that far back.

If this is an issue which has a big impact on you, we can do one of two things:
1) Investigate a workaround or special version tailored to your needs (and we'll have to discuss what those are, of course)
OR
2) If you have some time flexibility, I can make the refactoring a top priority (and you'll have to live with the issue for awhile).

**Comment by bgyori** [comment]
Thanks for the suggestions! Based on what I know has changed, it wouldn't be a huge problem to drop back to v1.3.4, which is from July. However, I've been having trouble building that version, with sbt assembly erroring due to a merge conflict for a dependency. If you prefer I can edit this comment and spin this off into a separate issue.

```
[error] 2 errors were encountered during merge
java.lang.RuntimeException: deduplicate: different file contents found in the following:
/home/bmg16/.ivy2/cache/org.maltparser/maltparser/jars/maltparser-1.9.0.jar:appdata/options.xml
/home/bmg16/.ivy2/cache/org.clulab/processors-models_2.11/jars/processors-models_2.11-6.0.7.jar:appdata/options.xml
deduplicate: different file contents found in the following:
/home/bmg16/.ivy2/cache/org.maltparser/maltparser/jars/maltparser-1.9.0.jar:appdata/release.properties
/home/bmg16/.ivy2/cache/org.clulab/processors-models_2.11/jars/processors-models_2.11-6.0.7.jar:appdata/release.properties
	at sbtassembly.Assembly$.applyStrategies(Assembly.scala:140)
	at sbtassembly.Assembly$.x$1$lzycompute$1(Assembly.scala:25)
	at sbtassembly.Assembly$.x$1$1(Assembly.scala:23)
	at sbtassembly.Assembly$.stratMapping$lzycompute$1(Assembly.scala:23)
	at sbtassembly.Assembly$.stratMapping$1(Assembly.scala:23)
	at sbtassembly.Assembly$.inputs$lzycompute$1(Assembly.scala:67)
	at sbtassembly.Assembly$.inputs$1(Assembly.scala:57)
	at sbtassembly.Assembly$.apply(Assembly.scala:83)
	at sbtassembly.Assembly$$anonfun$assemblyTask$1.apply(Assembly.scala:240)
	at sbtassembly.Assembly$$anonfun$assemblyTask$1.apply(Assembly.scala:237)
	at scala.Function1$$anonfun$compose$1.apply(Function1.scala:47)
	at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:40)
	at sbt.std.Transform$$anon$4.work(System.scala:63)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
	at sbt.Execute.work(Execute.scala:237)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
	at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
[error] (root/*:assembly) deduplicate: different file contents found in the following:
[error] /home/bmg16/.ivy2/cache/org.maltparser/maltparser/jars/maltparser-1.9.0.jar:appdata/options.xml
[error] /home/bmg16/.ivy2/cache/org.clulab/processors-models_2.11/jars/processors-models_2.11-6.0.7.jar:appdata/options.xml
[error] deduplicate: different file contents found in the following:
[error] /home/bmg16/.ivy2/cache/org.maltparser/maltparser/jars/maltparser-1.9.0.jar:appdata/release.properties
[error] /home/bmg16/.ivy2/cache/org.clulab/processors-models_2.11/jars/processors-models_2.11-6.0.7.jar:appdata/release.properties
[error] Total time: 71 s, completed Sep 26, 2017 4:39:21 PM
```

**Comment by hickst** [comment]
This has to do with some resources internal to the Malt parser whose version has to match the version of Malt used. Since you are reverting Reach you have to revert (and locally compile) Processors. In reviewing the checkins, I believe you have to revert Processors to version 97f4f79e1bc022cee38f733dcc3e3916a8ce4f81 or earlier.

But I still think this is opening a can of worms. Unless you have absolutely no time to wait on this issue, I would recommend that we discuss what I can do for you from this end. Feel free to call me directly and we can discuss the schedule and parameters of a solution.

**Comment by bgyori** [comment]
I managed to make a custom build that includes the latest Bioresources (important for us) and builds off of REACH 1.3.3 (which is fine for now). We'll keep monitoring #547 and update to the latest REACH once there is a fix for that. Thanks for your help!
